### PR TITLE
Add functions to get compatible QoS profiles

### PR DESCRIPTION
**Pull request by jacobperron**
Connects to https://github.com/ros2/rmw/issues/304

Currently, just contains an implementation for the subscription. I'll wait for feedback before adding the equivalent function for publishers.

### CODE REVIEW

**Review by wjwwood — COMMENTED** [review]
I think this is reasonable.

File: rmw_dds_common/src/qos.cpp [hunk 1]
```diff
@@ -374,4 +374,62 @@ qos_profile_check_compatible(
   return RMW_RET_OK;
 }
 
+rmw_ret_t
+qos_profile_get_most_compatible_for_subscription(
```
Comment (ivanpauno) [inline rmw_dds_common/src/qos.cpp:378]: should this function also take in accound `deadline`?

You need `deadline_publisher < deadline_subscription` for them to match

File: rmw_dds_common/include/rmw_dds_common/qos.hpp [hunk 2]
```diff
@@ -56,6 +57,80 @@ qos_profile_check_compatible(
...
+  rcutils_allocator_t * allocator);
+
+/// Get compatible QoS policies for a publisher.
+/**
+ * Given one or more subscription endpoints, return a QoS profile for a publisher
+ * that is compatible with the majority of the subscriptions while maintaining the highest
+ * level of service possible.
+ *
+ * This implements the rmw API \ref rmw_qos_profile_get_most_compatible_for_subscription().
+ * See \ref rmw_qos_profile_get_most_compatible_for_subscription() for more information.
```
Comment (fujitatomoya) [inline rmw_dds_common/include/rmw_dds_common/qos.hpp:104]:
```suggestion
 * This implements the rmw API \ref rmw_qos_profile_get_most_compatible_for_publisher().
 * See \ref rmw_qos_profile_get_most_compatible_for_publisher() for more information.
```

File: rmw_dds_common/src/qos.cpp [hunk 3]
```diff
@@ -374,4 +374,197 @@ qos_profile_check_compatible(
...
+
+  const rmw_time_t deadline_default = RMW_QOS_DEADLINE_DEFAULT;
+  const rmw_time_t liveliness_lease_duration_default = RMW_QOS_LIVELINESS_LEASE_DURATION_DEFAULT;
+
+  // Always use "reliable" reliability and "transient_local" durability since both policies
+  // are compatible with all subscriptions and have highest level of service
+  publisher_profile->reliability = RMW_QOS_POLICY_RELIABILITY_RELIABLE;
+  publisher_profile->durability = RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL;
+
+  // Only use "manual by topic" liveliness if at least one  subscription is using manual by topic
```
Comment (fujitatomoya) [inline rmw_dds_common/src/qos.cpp:518]:
```suggestion
  // Only use "manual by topic" liveliness if at least one subscription is using manual by topic
```

**Review by fujitatomoya — COMMENTED** [review]
a few minor nitpicks, it looks good to me.

**Comment by jacobperron** [comment]
Closing in favor of ros2/rmw#320

**Pull request closed without merging**